add image size to onLoad nativeEvent

dubert opened the pull request:
Issue: https://github.com/facebook/react-native/issues/494
Related: https://github.com/facebook/react-native/issues/2180

paramaggarwal commented:
Excellent. :)

dubert commented:
@paramaggarwal can you tell me why the checks failed? This is my first PR...

paramaggarwal commented:
Network failed on the CI server. Try pushing another commit to trigger a new build.

ide commented:
I think this returns the displayed size, not the intrinsic size. Can you verify?

hayeah commented:
@ide @dubert RCTImageView gets the resized dimensions.

Would need to change ` downloadImageForURL` to return the intrinsic image size.

Relevant code:

https://github.com/facebook/react-native/blob/f0dd9fb358efb94fba4dd2ba74fb4910bd929056/Libraries/Image/RCTImageLoader.m#L220

```
return [_bridge.imageDownloader downloadImageForURL:url size:size scale:scale resizeMode:resizeMode progressBlock:progressBlock completionBlock:^(NSError *error, id image) {
      RCTDispatchCallbackOnMainQueue(completionBlock, error, image);
    }];
```

dubert commented:
I don't know Objective-C, so here's my best shot. I expect there will be a number of revisions necessary to bring it up to snuff.

I believe there is value in providing both the displayed size and intrinsic image size. So I kept `size` and added `dimensions` and `scale` (so that pixel value can be calculated if wanted).

Because I modified the imageLoader function and callback, I had to update it in a bunch of locations, passing `nil` where dimensions weren't relevant. I didn't know a better way to approach this.

Thoughts @ide, @hayeah ?

paramaggarwal commented:
Wow, I didn't know that RN resizes the image based on the size it will be used at. Quite a nice memory optimisation. My vote is that only `dimensions` are needed as `size` is available on `onLayout` and scale is available using `PixelRatio`.

ide commented:
I'm not sure it's worth showing the displayed size because that should be the size of the Image component. Basically what @paramaggarwal said.

dubert commented:
Updated to only show size for network images.

paramaggarwal commented:
Looks good! Small nitpick - no need for the `if (intrinsicSize.height != 0 && intrinsicSize.width != 0)` - if it `0` it will be sent as `0` - not a problem.

dubert commented:
@paramaggarwal I'm checking it so that I don't pass `size: { height: 0, width: 0 }`. It's only the target.

paramaggarwal commented:
@dubert I don't think that is a problem. Let's wait for others' thoughts.

dubert commented:
pinging @ide @hayeah who needs to review this?

ide commented on Libraries/Image/RCTImageView.m:
```diff
@@ -126,7 +126,7 @@ - (void)reloadImage
                                resizeMode:self.contentMode
                             progressBlock:progressHandler
-                          completionBlock:^(NSError *error, id image) {
+                          completionBlock:^(NSError *error, id image, id dimensions) {
```
NSValue instead of id

ide commented on Libraries/Image/RCTImageView.m:
```diff
@@ -144,7 +144,19 @@ - (void)reloadImage
       } else {
         if (_onLoad) {
-          NSDictionary *event = @{ @"target": self.reactTag };
+          CGSize intrinsicSize = [dimensions CGSizeValue];
```
if dimensions is nil, then exclude the "size" from the event object (otherwise include it). This means that if we download an image that actually has 0 width or height, it will be reported to the JS code.

ide commented on Libraries/Image/RCTImageView.m:
```diff
@@ -144,7 +144,19 @@ - (void)reloadImage
         if (_onLoad) {
-          NSDictionary *event = @{ @"target": self.reactTag };
+          CGSize intrinsicSize = [dimensions CGSizeValue];
+          NSDictionary *event;
```
In terms of code style it might be more succinct to start with 
```
[NSMutableDictionary dictionaryWithDictionary:@{
  @"target": self.reactTag,
}];
```
and add "size" to it later

dubert commented:
@ide updated and squashed

jnhuynh commented:
+1

ide commented:
cc @a2 @nicklockwood for review

dubert commented:
@paramaggarwal any idea why the checks have failed? You were able to help me last time

paramaggarwal commented:
@dubert 
```
RCTImageLoaderTests.m:46:33: error: too few arguments to block call, expected 3, have 2
    completionHandler(nil, image);
```

dubert commented:
@ide, @a2, @nicklockwood for review? :smile_cat:

ide commented on Libraries/Image/RCTImageView.m:
```diff
@@ -194,7 +194,18 @@ - (void)reloadImage
         if (_onLoad) {
-          _onLoad(nil);
+          if (dimensions != nil) {
+            CGSize size = [dimensions CGSizeValue];
```
Could we greatly simplify this PR by using `image.size` (may need to be careful to account for screen scale -- but it sure would make this PR super short :D )?

nicklockwood commented:
I'm not sure I understand the purpose of this. Why not just get the size from the image directly? (image.size). Why do you care how big the image was originally before we resized it? How would this information be used?

nicklockwood commented:
OK, sorry, should've read the attached issue properly. I understand the problem, but this is the wrong solution. Loading the image at the wrong size using the <Image> tag and then resizing it to the size you want is pretty hacky, and I don't want to build an API around supporting a hack - if we're going to make API changes then we should support the use case properly, even if it means a bigger change.

I'd be fine with sending the *actual* loaded image size in the callback. One option would be to do that, then you could load the image at the maximum possible size and then resize it. Still a hack, but at least the API would be logical. You could embed the <Image> inside a smaller view with overflow: hidden, or obscure it behind another view while it's loading to prevent it from messing up the UI before it's resized.

In terms of solving the actual use case, I think we need to add a way to preload images and get their size without having to display the image while its loading. It would be relative easy to expose something like that, but I'll need to discuss with the Android team what the API should look like.

eyaleizenberg commented:
I tried adding these on RN 0.13 and I couldn't get the project to start... :cry:

olofd commented:
I really need this fix as well. Is there any work going on here?

dubert commented:
@olofd Nope. I don't know how to proceed. It sounds like the FB devs don't like the way I solved this. But it's not a feature they are planning to provide themselves. So right now I'm just having to use a forked version of react-native. I'm several versions behind because I'm tired of rebasing :(

olofd commented:
Ok. for everyone who has a problem with this and want to be able to support remote images with unknown dimensions I made a hack you can use before the react team supports the use-case.

This branch is not suppose to be beautiful, rather to touch as little as possible, so it'll be easy to keep merging, and still give you the functionality. 

It's based on 0.16.0:
https://github.com/olofd/react-native/tree/Remote-Image-Dimensions

Compare against 0.16.0:
https://github.com/facebook/react-native/compare/0.16-stable...olofd:Remote-Image-Dimensions?expand=1

To use it. Just change your package.json react-native-version to:
"react-native": "git://github.com/olofd/react-native.git#Remote-Image-Dimensions"

I'll try to keep merging it with new releases.

dubert commented:
That looks great @olofd, I'll use it in my project.

@nicklockwood Do you have any more thoughts about this PR? Should I go ahead and close it? Or is it worth trying to figure out how to preload images and get their size? Please keep in mind I have very little Objective-C experience.

nicklockwood commented:
I'm still actively investigating the right API for preloading. I don't want to add something we won't be able to support moving forward. Sorry about the wait.

olofd commented:
Updated my branch with this feature for v0.17.0-rc:
https://github.com/olofd/react-native/tree/Image-Dimensions-0.17.0-rc

Use this in package.json:
"react-native": "git://github.com/olofd/react-native.git#Image-Dimensions-0.17.0-rc"

seb0zz commented:
@olofd using your branch, too. Works like a charm :+1:

seb0zz commented:
@olofd do you have any experience on using this with Android? Seems like the events are not fired there (#4807)

olofd commented:
Updated my branch with this feature for v0.17.0:
https://github.com/olofd/react-native/tree/Image-Dimensions-0.17.0

Use this in package.json:
"react-native": "git://github.com/olofd/react-native.git#Image-Dimensions-0.17.0"

dbuarque commented:
So, what's the deal?
